src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,282 @@
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Performs an read of <see cref="_instance"/> and <see cref="_boxedReferenceCount"/>, where the mutation
+            /// performed by <see cref="Dispose"/> behaves as an atomic operation.
+            /// </summary>
+            /// <returns>
+            /// <para>If the returned reference count box is non-<see langword="null"/>, then the target will be set to
+            /// the object this reference was initialized with. However, the target will only be valid prior to the
+            /// reference count being set to zero.</para>
+            ///
+            /// <para>If the returned reference count box is <see langword="null"/>, then the target will also be
+            /// <see langword="null"/>.</para>
+            /// </returns>
+            private (T target, StrongBox<int> referenceCount) AtomicReadState()
+            {
+                // This value can be set in Dispose. A memory barrier is used to ensure _boxedReferenceCount is read
+                // after _instance (a stale read of _instance is not an error).
+                var target = _instance;
+                Interlocked.MemoryBarrier();
+                var referenceCount = Volatile.Read(ref _boxedReferenceCount);
+                if (referenceCount == null)
+                {
+                    return (null, null);
+                }
+                else
+                {
+                    return (target, referenceCount);
+                }
+            }
+
+            /// <summary>
+            /// Releases the current reference, causing the underlying object to be disposed if this was the last
+            /// reference.
+            /// </summary>
+            /// <remarks>
+            /// <para>After this instance is disposed, the <see cref="TryAddReference"/> method can no longer be used to
+            /// object a new reference to the target, even if other references to the target object are still in
+            /// use.</para>
+            /// </remarks>
+            public void Dispose()
+            {
+                var referenceCount = Interlocked.Exchange(ref _boxedReferenceCount, null);
+                if (referenceCount == null)
+                {
+                    // Already disposed; allow multiple without error.
+                    return;
+                }
+
+                // This is the thread which will write to _instance, so this will act as an atomic read
+                var instance = Interlocked.Exchange(ref _instance, null);
+
+                // Set the instance back to its default value, so in case someone forgets to dispose of one of the
+                // counted references the finalizer of the underlying disposable object will have a chance to clean up
+                // the unmanaged resources as soon as possible.
+                Interlocked.MemoryBarrier();
+                _instance = default(T);
```

Comment:
These lines were necessary prior to `T` having the generic constraint `class`, but are now a NOP. I have removed them now and updated the documentation for both this and the `Target` property.